src/ray/core_worker/reference_count.cc

```diff
@@ -198,10 +198,56 @@ void ReferenceCounter::AddOwnedObject(const ObjectID &object_id,
                                       bool is_reconstructable,
                                       bool add_local_ref,
                                       const absl::optional<NodeID> &pinned_at_raylet_id) {
-  RAY_LOG(DEBUG) << "Adding owned object " << object_id;
   absl::MutexLock lock(&mutex_);
-  RAY_CHECK(object_id_refs_.count(object_id) == 0)
+  RAY_CHECK(AddOwnedObjectInternal(object_id,
+                                   inner_ids,
+                                   owner_address,
+                                   call_site,
+                                   object_size,
+                                   is_reconstructable,
+                                   add_local_ref,
+                                   pinned_at_raylet_id))
       << "Tried to create an owned object that already exists: " << object_id;
+}
+
+void ReferenceCounter::AddDynamicReturn(const ObjectID &object_id,
+                                        const ObjectID &generator_id) {
+  absl::MutexLock lock(&mutex_);
+  auto outer_it = object_id_refs_.find(generator_id);
+  if (outer_it == object_id_refs_.end()) {
+    // Outer object already went out of scope before we could deserialize the
+    // inner one, so there can't be references to the inner one anymore.
```

Comment:
Add a check to make sure?

Reply:
Actually I don't think the comment is accurate. Either the outer object went out of scope already without deserializing the inner one, so we don't need to add the dynamic ref. Or the outer object went out of scope after deserializing the inner one, so the dynamic ref is already added. Will update.